notebook/static/edit/js/editor.js

```diff
@@ -200,32 +212,115 @@ function(
             function (model) {
                 that.file_path = model.path;
                 that.events.trigger('file_renamed.Editor', model);
+                that.last_modified = new Date(model.last_modified);
                 that._set_mode_for_model(model);
                 that._clean_state();
             }
         );
     };
     
-    Editor.prototype.save = function () {
+
+    /**
+     * Save this file on the server.
+     *
+     * @param {boolean} check_last_modified - checks if file has been modified on disk
+     * @return {Promise} - promise that resolves when the notebook is saved.
+     */
+    Editor.prototype.save = function (check_last_modified) {
         /** save the file */
         if (!this.save_enabled) {
             console.log("Not saving, save disabled");
             return;
         }
+
+        // used to check for last modified saves
+        if (check_last_modified === undefined) {
+            check_last_modified = true;
+        }
+
         var model = {
             path: this.file_path,
             type: 'file',
             format: 'text',
             content: this.codemirror.getValue(),
         };
         var that = this;
+
         // record change generation for isClean
+        // (I don't know what this implies for the editor)
```

Comment:
The 'generation' is a number Codemirror uses to tell whether the editor is 'clean' - whether the current text matches what is saved. This gets the current generation when we save the file. Elsewhere in the code, we pass the number to `isClean` to ask Codemirror if the file has changed since the last save.

This means that if you save, type something, and Ctrl-z to undo, we know that you're back at the saved state.

So this should probably only happen when we're actually about to save. If the users cancels the dialog, `this.generation` shouldn't have changed.

Reply:
Ah, fascinating! Thanks for the explanation and link. I'll also pull the changeGeneration method within the promise 'then' statement that validates the save.